src/utils.rs

```diff
@@ -35,6 +37,215 @@ use crate::FrameLayout;
+            curr_byte: Default::default(),
+            num_remaining_bits_in_curr_byte: Default::default(),
+            prev_two_bytes: 0xffff,
+            num_epb: Default::default(),
+            needs_epb: needs_epb,
+        }
+    }
+
+    /// Read a single bit from the stream.
+    pub fn read_bit(&mut self) -> Result<bool, ReadBitsError> {
+        let bit = self.read_bits::<u32>(1)?;
+        match bit {
+            1 => Ok(true),
+            0 => Ok(false),
+            _ => panic!("Unexpected value {}", bit),
+        }
+    }
+
+    /// Read up to 31 bits from the stream.
+    pub fn read_bits<U: TryFrom<u32>>(&mut self, num_bits: usize) -> Result<U, ReadBitsError> {
+        if num_bits > 31 {
+            return Err(ReadBitsError::TooManyBytesRequested(num_bits));
+        }
+
+        let mut bits_left = num_bits;
+        let mut out = 0u32;
+
+        while self.num_remaining_bits_in_curr_byte < bits_left {
+            out |= (self.curr_byte as u32) << (bits_left - self.num_remaining_bits_in_curr_byte);
+            bits_left -= self.num_remaining_bits_in_curr_byte;
+            self.update_curr_byte()?;
+        }
+
+        out |= (self.curr_byte >> (self.num_remaining_bits_in_curr_byte - bits_left)) as u32;
+        out &= (1 << num_bits) - 1;
+        self.num_remaining_bits_in_curr_byte -= bits_left;
+
+        U::try_from(out).map_err(|_| ReadBitsError::ConversionFailed)
+    }
+
+    /// Skip `num_bits` bits from the stream.
+    pub fn skip_bits(&mut self, mut num_bits: usize) -> Result<(), ReadBitsError> {
+        while num_bits > 0 {
+            let n = std::cmp::min(num_bits, 31);
+            self.read_bits::<u32>(n)?;
+            num_bits -= n;
+        }
+
+        Ok(())
+    }
+
+    /// Returns the amount of bits left in the stream
+    pub fn num_bits_left(&self) -> usize {
+        self.data.remaining() * 8 + self.num_remaining_bits_in_curr_byte
+    }
+
+    /// Returns the number of emulation-prevention bytes read so far.
+    pub fn num_epb(&self) -> usize {
+        self.num_epb
+    }
+
+    /// Whether the stream still has RBSP data. Implements more_rbsp_data(). See
+    /// the spec for more details.
+    pub fn has_more_rsbp_data(&mut self) -> bool {
+        if self.num_remaining_bits_in_curr_byte == 0 && self.update_curr_byte().is_err() {
+            // no more data at all in the rbsp
+            return false;
+        }
+
+        // If the next bit is the stop bit, then we should only see unset bits
+        // until the end of the data.
+        if (self.curr_byte & ((1 << (self.num_remaining_bits_in_curr_byte - 1)) - 1)) != 0 {
+            return true;
+        }
+
+        let data = self.data.chunk();
+        for data in &data[0..self.data.remaining()] {
+            if *data != 0 {
+                return true;
+            }
+        }
+
+        self.data.advance(self.data.remaining());
+        false
+    }
+
+    pub fn read_ue<U: TryFrom<u32>>(&mut self) -> anyhow::Result<U> {
```

Comment:
Nit: please add a doc comment.

I.e.

```
/// Reads an Unsigned Exponential golomb coding number from the next bytes in the
/// bitstream. This may advance the state of position within the bitstream even if the
/// read operation is unsuccessful.
```

Or something like that. WDYT?

Reply:
Similarly for other `read*` or `get*`  methods.